hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java

```diff
@@ -0,0 +1,303 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hdds.scm.security;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.scm.ha.SCMContext;
+import org.apache.hadoop.hdds.scm.ha.SCMService;
+import org.apache.hadoop.hdds.scm.ha.SCMServiceException;
+import org.apache.hadoop.hdds.scm.server.StorageContainerManager;
+import org.apache.hadoop.hdds.security.x509.certificate.client.CertificateClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.cert.X509Certificate;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_X509_CA_ROTATION_CHECK_INTERNAL;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_X509_CA_ROTATION_CHECK_INTERNAL_DEFAULT;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_X509_CA_ROTATION_TIME_OF_DAY;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_X509_CA_ROTATION_TIME_OF_DAY_DEFAULT;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_X509_RENEW_GRACE_DURATION;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_X509_RENEW_GRACE_DURATION_DEFAULT;
+
+/**
+ * Root CA Rotation Manager is a service in SCM to control the CA rotation.
+ */
+public class RootCARotationManager implements SCMService {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(RootCARotationManager.class);
+
+  private StorageContainerManager scm;
+  private final SCMContext scmContext;
+  private OzoneConfiguration ozoneConf;
+  private ScheduledExecutorService executorService;
+  private Duration checkInterval;
+  private Duration renewalGracePeriod;
+  private Pattern timePattern = Pattern.compile("\\d{2}:\\d{2}:\\d{2}");
+  private Date timeOfDay;
+  private CertificateClient scmCertClient;
+  private AtomicBoolean isRunning = new AtomicBoolean(false);
+  private AtomicBoolean isScheduled = new AtomicBoolean(false);
+  private String threadName = this.getClass().getSimpleName() + "-MonitorTask";
+
+  /**
+   * Constructs RootCARotationManager with the specified arguments.
+   *
+   * @param scm the storage container manager
+   */
+  public RootCARotationManager(StorageContainerManager scm) {
+    this.scm = scm;
+    this.ozoneConf = scm.getConfiguration();
+
+    this.scmContext = scm.getScmContext();
+    String durationString = ozoneConf.get(
```

Comment:
Instead of relying on OzoneConfiguration, wouldn't it be better to add these config processing to the SecurityConfig object, and get the values from there with the properly typed values?
I think these we might need in other services as well in order to determine the time when other services should start to expect a new rootCA certificate, so that seems to be a natural common place for getting these values.

Reply:
Good point.